api/tests/unit/segments/test_unit_segments_models.py

```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
 )
 def test_segment_id_exists_in_rules_data(rules_data, expected_result):
     assert Segment.id_exists_in_rules_data(rules_data) == expected_result
+
+
+def test_deep_clone_of_segment(
+    project: Project,
+    feature: Feature,
+) -> None:
+    # Given
+    segment = Segment.objects.create(
+        name="SpecialSegment",
+        description="A lovely, special segment.",
+        project=project,
+        feature=feature,
+    )
+
+    # Check that the versioning is correct, since we'll testing
+    # against it later in the test.
+    assert segment.version == 1
+    assert segment.version_of == segment
+
+    parent_rule1 = SegmentRule.objects.create(
+        segment=segment, type=SegmentRule.ALL_RULE
+    )
+    parent_rule2 = SegmentRule.objects.create(
+        segment=segment, type=SegmentRule.ANY_RULE
+    )
+
+    child_rule1 = SegmentRule.objects.create(
+        rule=parent_rule1, type=SegmentRule.ANY_RULE
+    )
+    child_rule2 = SegmentRule.objects.create(
+        rule=parent_rule1, type=SegmentRule.ALL_RULE
+    )
+    child_rule3 = SegmentRule.objects.create(
+        rule=parent_rule2, type=SegmentRule.NONE_RULE
+    )
+    child_rule4 = SegmentRule.objects.create(
+        rule=parent_rule2, type=SegmentRule.ALL_RULE
+    )
+
+    condition1 = Condition.objects.create(
+        rule=parent_rule1,
+        property="parent_rule1",
+        operator=EQUAL,
+        value="condition1",
+        created_with_segment=True,
+    )
+    condition2 = Condition.objects.create(
+        rule=parent_rule1,
+        property="parent_rule1",
+        operator=PERCENTAGE_SPLIT,
+        value="0.1",
+        created_with_segment=False,
+    )
+
+    condition3 = Condition.objects.create(
+        rule=child_rule1,
+        property="child_rule1",
+        operator=EQUAL,
+        value="condition3",
+        created_with_segment=True,
+    )
+    Condition.objects.create(
+        rule=child_rule2,
+        property="child_rule2",
+        operator=PERCENTAGE_SPLIT,
+        value="0.2",
+        created_with_segment=False,
+    )
+    Condition.objects.create(
+        rule=child_rule2,
+        property="child_rule2",
+        operator=EQUAL,
+        value="condition5",
+        created_with_segment=False,
+    )
+
+    Condition.objects.create(
+        rule=child_rule3,
+        property="child_rule3",
+        operator=EQUAL,
+        value="condition6",
+        created_with_segment=False,
+    )
+
+    Condition.objects.create(
+        rule=child_rule4,
+        property="child_rule4",
+        operator=EQUAL,
+        value="condition7",
+        created_with_segment=True,
+    )
+
+    # When
+    new_segment = segment.deep_clone()
```

Comment:
I think we could perhaps improve the naming here and use `cloned_segment` instead of `new_segment` ?